### ISSUE TITLE
Interrupting chunked-encoded request may affect with looping one thread and consuming 100% of CPU

### ISSUE THREAD

**Issue by ddzialak** [benoitc/gunicorn#2684]
```
Thread 406626 (active): "gthread-406239_22"
    chunk (gunicorn/http/unreader.py:65)
    read (gunicorn/http/unreader.py:38)
    parse_chunked (gunicorn/http/body.py:68)
    read (gunicorn/http/body.py:28)
    read (gunicorn/http/body.py:212)
    stream_request_body (sfutils/service/base_service.py:842)
```

and [gunicorn/http/body.py in line 69](https://github.com/benoitc/gunicorn/blob/master/gunicorn/http/body.py#L69) (my copy of gunicorn isn't newest one so lines are a bit shifted) has a function `ChunkedReader.parse_chunked()`:
``` 
    def parse_chunked(self, unreader):
        (size, rest) = self.parse_chunk_size(unreader)
        while size > 0:
            while size > len(rest):
                size -= len(rest)
                yield rest
                rest = unreader.read()
                if not rest:
                    raise NoMoreData()
            yield rest[:size]
            # Remove \r\n after chunk
            rest = rest[size:]
            while len(rest) < 2:
                rest += unreader.read()
            if rest[:2] != b'\r\n':
                raise ChunkMissingTerminator(rest[:2])
            (size, rest) = self.parse_chunk_size(unreader, data=rest[2:])
```
So, my guess is that: if chunked-encoded request will handdle connection, and will read whole chunk properly, but connection will be interrupted before sending `\r\n` then gunicorn thread will stuck in infinite loop:
```
            while len(rest) < 2:
                rest += unreader.read()
```
so using exception from the previous loop here:
```
            while len(rest) < 2:
                new_data = unreader.read()
                if not new_data:
                    raise NoMoreData()
                rest += new_data
```
 should fix the problem...

**Comment by forrestrice** [comment]
Unfortunately I have not been able to build an independent way of reproducing this, but we have a service behind a proxy/load-balancer that is applying chunked transfer encoding to large payloads and we were seeing cases where worker threads would stop processing new requests and high CPU usage was reported. Eventually all worker threads would get in that state and the service would be totally unresponsive. I was able to reproduce it in our environments via load tests against the load-balancer but not locally with chunked load tests against the local service, so it could be the case that either the load-balancer or the production environment is causing an interruption and triggering this issue, but I cannot prove it.

For the time being we have deployed nginx in front of our service to buffer incoming chunked requests and we have not observed the problem since.
